pkg/gator/reader/read_constraints.go

```diff
@@ -207,3 +226,51 @@ func ReadK8sResources(r io.Reader) ([]*unstructured.Unstructured, error) {
 
 	return objs, nil
 }
+
+// ReadSyncRequirements parses the sync requirements from a
+// constraint template.
+func ReadSyncRequirements(t *templates.ConstraintTemplate) (*SyncRequirements, error) {
+	syncRequirements := &SyncRequirements{}
+	if t.ObjectMeta.Annotations != nil {
+		if annotation, exists := t.ObjectMeta.Annotations[SyncAnnotationName]; exists {
+			annotation = strings.Trim(annotation, "\n\"")
+			err := json.Unmarshal([]byte(annotation), syncRequirements)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return syncRequirements, nil
+}
+
+func (r *SyncRequirements) UnmarshalJSON(data []byte) error {
```

Comment:
It may be easier for people to grok the code if we don't rely on magic methods from the JSON package.

In other words `ReadSyncRequirements` unmarshalls the JSON directly into `[][]compactGVKEquivalentSet`, then does the conversion itself.

Reply:
Doing this also makes the object round-trippable: right now, I believe Marshaling/Unmarshaling a SyncRequirements object would fail, since the serialized form would be different (you'd need a custom marshaler to fix it otherwise, and at that point there'd be information loss -- the GVKs may not be compacted in the same way)

Reply:
Thanks! Changed

Reply:
Actually just remembered the question I had here was that assuming this is a library other pieces of code use--if someone has their own annotation they want to unmarshal we want to give them a way to do so without having to provide a template that follows our struct--maybe that's just adding one more layer of wrappers or just changing this to accept the annotation string only--lmk your thoughts

Reply:
I don't follow. Are you asking about if someone wants to provide an arbitrary annotation? An annotation that describes the same thing (sync requirements) but using a different annotation key?

Reply:
Basically if they wanted to use our unmarshalling logic in the original format, they could pass any slice of bytes and unmarshal it to our SyncRequirements kind without having to pass a whole ConstraintTemplate struct, vs now so much of the code lives in ReadAnnotation which takes a template

Reply:
I see. I'd probably:

* Make the `compactGVKEquivalenceSet` struct public
* Have a public function that converts `compactGVKEquivalenceSet` to the denormalized version
* Keep this same function that translates a template into this requirements (for convenience and consistency of annotation names). This function would be built on top of the other two.

Basically this allows for using sources-of-truths other than annotation, and does not assume the input is serialized JSON (though it's easy enough for users to coerce serialized JSON into `compactGVKEquivalenceSet` if necessary).

Reply:
That makes a lot of sense! Made that change, thanks!